hamilton/driver.py

```diff
@@ -472,38 +546,79 @@ def raw_execute(
         """
-        nodes, user_nodes = self.graph.get_upstream_nodes(final_vars, inputs, overrides)
+        function_graph = _fn_graph if _fn_graph is not None else self.graph
+        run_id = str(uuid.uuid4())
+        nodes, user_nodes = function_graph.get_upstream_nodes(final_vars, inputs, overrides)
         Driver.validate_inputs(
-            self.graph, self.adapter, user_nodes, inputs, nodes
+            function_graph, self.adapter, user_nodes, inputs, nodes
         )  # TODO -- validate within the function graph itself
         if display_graph:  # deprecated flow.
             logger.warning(
                 "display_graph=True is deprecated. It will be removed in the 2.0.0 release. "
                 "Please use visualize_execution()."
             )
             self.visualize_execution(final_vars, "test-output/execute.gv", {"view": True})
-            if self.has_cycles(final_vars):  # here for backwards compatible driver behavior.
+            if self.has_cycles(
+                final_vars, function_graph
+            ):  # here for backwards compatible driver behavior.
                 raise ValueError("Error: cycles detected in your graph.")
         all_nodes = nodes | user_nodes
         self.graph_executor.validate(list(all_nodes))
-        return self.graph_executor.execute(
-            self.graph,
-            final_vars,
-            overrides if overrides is not None else {},
-            inputs if inputs is not None else {},
-        )
+        if self.adapter.does_hook("pre_graph_execute", is_async=False):
+            self.adapter.call_all_lifecycle_hooks_sync(
+                "pre_graph_execute",
+                run_id=run_id,
+                graph=function_graph,
+                final_vars=final_vars,
+                inputs=inputs,
+                overrides=overrides,
+            )
+        try:
+            out = self.graph_executor.execute(
+                function_graph,
+                final_vars,
+                overrides if overrides is not None else {},
+                inputs if inputs is not None else {},
+                run_id,
+            )
+            if self.adapter.does_hook("post_graph_execute", is_async=False):
+                self.adapter.call_all_lifecycle_hooks_sync(
+                    "post_graph_execute",
+                    run_id=run_id,
+                    graph=function_graph,
+                    success=True,
+                    error=None,
+                    results=out,
+                )
+        except Exception as e:
+            if self.adapter.does_hook("post_graph_execute", is_async=False):
+                self.adapter.call_all_lifecycle_hooks_sync(
+                    "post_graph_execute",
+                    run_id=run_id,
+                    graph=function_graph,
+                    success=False,
+                    error=e,
+                    results=None,
+                )
+            raise e
+
+        # TODO -- add error case
```

Comment:
what  error case needs to be caught here?

Reply:
This is done